crates/ra_hir/src/ty.rs

```diff
@@ -872,6 +873,155 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
+        };
+
+        for subpat in subpats {
+            let matching_field = fields.iter().find(|field| field.name == subpat.name);
+
+            if let Some(field) = matching_field {
+                let typeref = &field.type_ref;
+                let sub_ty = self.make_ty(typeref);
+                self.infer_pat(subpat.pat, &Expectation::has_type(sub_ty));
+            }
+        }
+
+        ty
+    }
+
+    // TODO: Expectation should probably contain a Cow pointer to Ty?
+    // so that we can make new expectations of subtypes cheaply
+    fn infer_pat(&mut self, pat: PatId, expected: &Expectation) -> Ty {
+        let body = Arc::clone(&self.body); // avoid borrow checker problem
+
+        let ty = match &body[pat] {
+            Pat::Tuple(ref args) => {
+                // this can probably be done without cloning/ collecting
+                let expectations = match expected.ty {
+                    Ty::Tuple(ref tuple_args) if args.len() == tuple_args.len() => {
+                        tuple_args.iter().cloned().collect()
+                    }
+                    _ => vec![Ty::Unknown; args.len()],
+                };
+
+                let inner_tys = args
+                    .iter()
+                    .zip(expectations.into_iter())
+                    .map(|(&pat, ty)| self.infer_pat(pat, &Expectation::has_type(ty)))
+                    .collect::<Vec<_>>()
+                    .into();
+
+                Ty::Tuple(inner_tys)
+            }
+            Pat::Ref { pat, mutability } => {
+                let expectation = match expected.ty {
+                    Ty::Ref(ref sub_ty, exp_mut) if *mutability == exp_mut => {
+                        Expectation::has_type((&**sub_ty).clone())
+                    }
+                    _ => Expectation::none(),
+                };
+                let subty = self.infer_pat(*pat, &expectation);
+                Ty::Ref(subty.into(), *mutability)
+            }
+            Pat::TupleStruct {
+                path: ref p,
+                args: ref subpats,
+            } => self.infer_tuple_struct(p.as_ref(), subpats),
+            Pat::Struct {
+                path: ref p,
+                args: ref fields,
+            } => self.infer_struct(p.as_ref(), fields),
+            Pat::Path(path) => {
+                // is this right?
+                self.module
+                    .resolve_path(self.db, &path)
+                    .take_values()
+                    .map_or(Ty::Unknown, |resolved| self.db.type_for_def(resolved))
+            }
+            Pat::Bind {
+                mode,
+                name: _name,
+                subpat,
+            } => {
+                let subty = if let Some(subpat) = subpat {
+                    self.infer_pat(*subpat, expected)
+                } else {
+                    let ty = self.new_type_var();
+                    self.unify(&ty, &expected.ty);
+                    let ty = self.resolve_ty_as_possible(ty);
```

Comment:
these three lines seem unnecessary? why not just use `expected.ty`?

(except maybe do `insert_type_vars_shallow`, since it may be `Unknown` and then get wrapped in `Ref`...)

Reply:
I tried that, but it doesn't actually 'bind' without explicitly resolving it.